### ISSUE TITLE
Ruby 2.6 segfault on macOS

### ISSUE THREAD

**Issue by brodygov** [aws/aws-sdk-ruby#2096]
### Issue description

Ruby segfault while iterating over AWS resource collection on Ruby 2.6 on macOS.

### Gem name:

aws-sdk-core 3.61.1 / aws-sdk-ec2 1.102.0

### Version of Ruby, OS environment:
Ruby 2.6 (tested 2.6.2 and 2.6.3), macOS Mojave (10.14.5)

Unaffected: Ruby 2.5 (tested 2.5.5, which works correctly)

### Code snippets / steps to reproduce

Code to reproduce:
```ruby
require 'aws-sdk-ec2'
r = Aws::EC2::Resource.new
puts r.vpcs.first
```

If you write `r.vpcs.to_a.first` instead of `r.vpcs.first`, the code works as expected.

I don't think there's anything special about the AWS account here. I tested against three different AWS accounts and got the same result.

Ruby prints the segfault information, then pins at 100% CPU. I killed it with a SIGKILL after 60 seconds, and it didn't appear to save the C level backtrace or anything in the crash reports folder.

Obviously this is probably a ruby bug. I haven't yet created a bug report there, since I was hoping for a more minimal reproducer but wasn't familiar with the aws-sdk-core internals.

```
$ timeout -k 60 4 ruby -r aws-sdk-ec2 -e 'r = Aws::EC2::Resource.new; puts r.vpcs.first'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:83: [BUG] Segmentation fault at 0x000070000cae3000
ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-darwin18]

-- Crash Report log information --------------------------------------------
   See Crash Report log file under the one of following:                    
     * ~/Library/Logs/DiagnosticReports                                     
     * /Library/Logs/DiagnosticReports                                      
   for more details.                                                        
Don't forget to include the above Crash Report log file in bug reports.     

-- Control frame information -----------------------------------------------
c:0043 p:---- s:0238 e:000237 CFUNC  :start
c:0042 p:0024 s:0234 E:001988 BLOCK  ~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:83
c:0041 p:0098 s:0228 E:000a50 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:103
c:0040 p:0092 s:0216 E:000878 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:945
c:0039 p:0004 s:0203 e:000202 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:930
c:0038 p:0048 s:0199 e:000198 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:925
c:0037 p:0040 s:0195 e:000194 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/delegate.rb:83
c:0036 p:0233 s:0186 e:000185 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/connection_p
c:0035 p:0032 s:0179 e:000178 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/connection_p
c:0034 p:0015 s:0172 e:000171 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/handler.rb:1
c:0033 p:0009 s:0165 e:000164 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/handler.rb:7
c:0032 p:0019 s:0157 e:000156 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/handler.rb:4
c:0031 p:0041 s:0152 e:000151 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/content_lengt
c:0030 p:0008 s:0146 e:000145 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/xml/error_handler.rb:8
c:0029 p:0021 s:0141 e:000140 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/signature_v4.rb:
c:0028 p:0035 s:0136 e:000135 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/transfer_encodin
c:0027 p:0008 s:0131 e:000130 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/helpful_socket_e
c:0026 p:0008 s:0125 e:000124 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/retry_errors.rb:
c:0025 p:0015 s:0119 e:000118 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/query/handler.rb:28
c:0024 p:0015 s:0114 e:000113 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/user_agent.rb:13
c:0023 p:0041 s:0109 e:000108 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/endpoint_pattern
c:0022 p:0133 s:0103 e:000102 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/endpoint_discove
c:0021 p:0038 s:0096 e:000095 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/endpoint.rb:4
c:0020 p:0034 s:0091 e:000090 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/param_validator.
c:0019 p:0008 s:0086 e:000085 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/raise_respons
c:0018 p:0028 s:0080 e:000079 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/jsonvalue_conver
c:0017 p:0026 s:0075 e:000074 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/idempotency_toke
c:0016 p:0046 s:0070 e:000069 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/param_converter.
c:0015 p:0021 s:0064 e:000063 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/response_paging.
c:0014 p:0073 s:0058 e:000057 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/response_targ
c:0013 p:0039 s:0052 e:000051 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/request.rb:70
c:0012 p:0025 s:0046 e:000045 METHOD ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-ec2-1.102.0/lib/aws-sdk-ec2/client.rb:21765
c:0011 p:0009 s:0039 e:000038 BLOCK  ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-ec2-1.102.0/lib/aws-sdk-ec2/resource.rb:3080 [FINISH]
c:0010 p:---- s:0034 E:002100 CFUNC  :each
c:0009 p:---- s:0031 e:000030 CFUNC  :each
c:0008 p:0008 s:0027 E:000558 BLOCK  ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb: [FINISH]
c:0007 p:---- s:0023 E:0007c0 CFUNC  :each
c:0006 p:---- s:0020 e:000019 CFUNC  :each
c:0005 p:0009 s:0016 E:000788 BLOCK  ~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb: [FINISH]
c:0004 p:---- s:0012 E:0026f0 CFUNC  :each
c:0003 p:---- s:0009 e:000008 CFUNC  :each
c:0002 p:---- s:0006 E:001800 IFUNC  
c:0001 p:---- s:0003 e:000002 (none) [FINISH]

-- Ruby level backtrace information ----------------------------------------
-e:0:in `each'
-e:0:in `each'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb:50:in `block in each'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb:50:in `each'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb:50:in `each'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb:99:in `block in non_empty_batches'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb:99:in `each'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/resources/collection.rb:99:in `each'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-ec2-1.102.0/lib/aws-sdk-ec2/resource.rb:3080:in `block in vpcs'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-ec2-1.102.0/lib/aws-sdk-ec2/client.rb:21765:in `describe_vpcs'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/request.rb:70:in `send_request'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/response_target.rb:23:in `call' 
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/response_paging.rb:10:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/param_converter.rb:24:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/idempotency_token.rb:17:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:20:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/raise_response_errors.rb:14:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/param_validator.rb:24:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/endpoint.rb:45:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/endpoint_discovery.rb:78:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/endpoint_pattern.rb:28:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/user_agent.rb:13:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/query/handler.rb:28:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/retry_errors.rb:174:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/helpful_socket_errors.rb:10:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/transfer_encoding.rb:26:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/plugins/signature_v4.rb:66:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/aws-sdk-core/xml/error_handler.rb:8:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/plugins/content_length.rb:17:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/handler.rb:47:in `call'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/handler.rb:73:in `transmit'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/handler.rb:121:in `session'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/connection_pool.rb:96:in `session_for'
~/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/aws-sdk-core-3.61.1/lib/seahorse/client/net_http/connection_pool.rb:297:in `start_session'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/delegate.rb:83:in `method_missing'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:925:in `start'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:945:in `connect'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:103:in `timeout'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:83:in `block in timeout'
~/.rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:83:in `start'

-- Machine register context ------------------------------------------------
 rax: 0x000070000cae3000 rbx: 0x0000000000103000 rcx: 0x000070000cae30ac
 rdx: 0x000070000c9e2000 rdi: 0x0000000000000002 rsi: 0x000070000cae3000
 rbp: 0x000000010b3d4680 rsp: 0x000000010b3d4640  r8: 0x00000000d1594ad4
  r9: 0x00000001080f6800 r10: 0x00007fff79731f4a r11: 0x0000000000000206
 r12: 0x00007fffaff6e60c r13: 0x0000000000100000 r14: 0x0000000000101000
 r15: 0x000000010b3d46f0 rip: 0x00007fff797ed0bb rfl: 0x0000000000010246

-- C level backtrace information -------------------------------------------
Killed: 9
```

**Comment by cjyclaire** [comment]
Thanks for the report!

I tried to reproduce with
```
ruby 2.6.2p47 (2019-03-13 revision 67232) [x86_64-darwin16]
```
on MacOs Sierra, but cannot, it might be that I have fewer VPCs? Mind sharing the approximate size for your VPCs? (Side note, I have tried `2.7.0-dev`, still not able to reproduce this.)

The `.vpcs` [itself](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-ec2/lib/aws-sdk-ec2/resource.rb#L3078), is basically a wrapper([`Aws::Resource::Collection`](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/resources/collection.rb)) over an enumerator. 

Based on the fact that it cannot be reproduced on every supported version, it could potentially be a Ruby-lang bug or SDK is not using best practices in newer language versions? If the later, happy to investigate more.
